dll/win32/kernel32/kernel32_vista/vista.c

```diff
@@ -26,61 +26,60 @@
+    _Inout_ PDWORD lpdwSize)
 {
-    BYTE Buffer[sizeof(UNICODE_STRING) + MAX_PATH * sizeof(WCHAR)];
-    UNICODE_STRING *DynamicBuffer = NULL;
-    UNICODE_STRING *Result = NULL;
+    BOOL bRet;
+    DWORD dwBufferSize;
+    PUNICODE_STRING pBuffer;
     NTSTATUS Status;
-    DWORD Needed;
+    DWORD dwCch;
 
-    Status = NtQueryInformationProcess(hProcess,
-                                       ProcessImageFileName,
-                                       Buffer,
-                                       sizeof(Buffer) - sizeof(WCHAR),
-                                       &Needed);
-    if (Status == STATUS_INFO_LENGTH_MISMATCH)
+    if (dwFlags & ~PROCESS_NAME_NATIVE)
     {
-        DynamicBuffer = RtlAllocateHeap(RtlGetProcessHeap(), 0, Needed + sizeof(WCHAR));
-        if (!DynamicBuffer)
-        {
-            BaseSetLastNTError(STATUS_NO_MEMORY);
-            return FALSE;
-        }
-
-        Status = NtQueryInformationProcess(hProcess,
-                                           ProcessImageFileName,
-                                           (LPBYTE)DynamicBuffer,
-                                           Needed,
-                                           &Needed);
-        Result = DynamicBuffer;
+        BaseSetLastNTError(STATUS_INVALID_PARAMETER_2);
+        return FALSE;
     }
-    else Result = (PUNICODE_STRING)Buffer;
-
-    if (!NT_SUCCESS(Status)) goto Cleanup;
 
-    if (Result->Length / sizeof(WCHAR) + 1 > *pdwSize)
+    dwBufferSize = sizeof(UNICODE_STRING) + *lpdwSize * sizeof(WCHAR);
+    pBuffer = (PUNICODE_STRING)RtlAllocateHeap(RtlGetProcessHeap(), 0, dwBufferSize);
+    if (!pBuffer)
     {
-        Status = STATUS_BUFFER_TOO_SMALL;
-        goto Cleanup;
+        BaseSetLastNTError(STATUS_NO_MEMORY);
+        return FALSE;
     }
 
-    *pdwSize = Result->Length / sizeof(WCHAR);
-    memcpy(lpExeName, Result->Buffer, Result->Length);
-    lpExeName[*pdwSize] = 0;
-
-Cleanup:
-    RtlFreeHeap(RtlGetProcessHeap(), 0, DynamicBuffer);
-
-    if (!NT_SUCCESS(Status))
+    bRet = FALSE;
+    Status = NtQueryInformationProcess(hProcess,
+                                       (dwFlags & PROCESS_NAME_NATIVE) ? ProcessImageFileName : ProcessImageFileNameWin32,
+                                       pBuffer,
+                                       dwBufferSize,
+                                       NULL);
+    if (NT_SUCCESS(Status))
     {
-        BaseSetLastNTError(Status);
+        dwCch = pBuffer->Length / sizeof(WCHAR);
+        if (dwCch >= *lpdwSize)
+        {
+            BaseSetLastNTError(STATUS_BUFFER_TOO_SMALL);
+        }
+        else
+        {
+            RtlCopyMemory(lpExeName, pBuffer->Buffer, pBuffer->Length);
+            lpExeName[dwCch] = UNICODE_NULL;
+            *lpdwSize = dwCch;
+            bRet = TRUE;
+        }
+    }
+    else
+    {
+        BaseSetLastNTError(Status == STATUS_INFO_LENGTH_MISMATCH ? STATUS_BUFFER_TOO_SMALL : Status);
```

Comment:
```suggestion
        if (Status == STATUS_INFO_LENGTH_MISMATCH)
            Status = STATUS_BUFFER_TOO_SMALL;
        BaseSetLastNTError(Status);
```
The same, but step-by-step-debugging friendly.